File: common/src/main/java/cc/turtl/chiselmon/feature/pc/PCUserData.java
package cc.turtl.chiselmon.feature.pc;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/**
 * Per world PC User Data.
 */
public class PCUserData {
    public final Bookmarks bookmarks = new Bookmarks();

    public static class Bookmarks {
        public static final int MAX_BOOKMARKS = 5;
        private final Set<Integer> data = new LinkedHashSet<>();

        public boolean toggle(int boxNo) {
            return data.remove(boxNo) || (data.size() < MAX_BOOKMARKS && data.add(boxNo));
        }

        public void remove(int boxNo) {
            data.remove(boxNo);
        }

        public void clear() {
            data.clear();
        }

        public boolean isFull() {
            return data.size() >= MAX_BOOKMARKS;
        }

        public boolean has(int boxNo) {
            return data.contains(boxNo);
        }

        public List<Integer> get() {
            return new ArrayList<>(data);
        }
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/feature/pc/PCUserData.kt
package cc.turtl.chiselmon.client.feature.pc

/**
 * Per world PC User Data.
 */
class PCUserData {
    val bookmarks = Bookmarks()

    class Bookmarks {
        companion object {
            const val MAX_BOOKMARKS = 5
        }

        private val data = LinkedHashSet<Int>()

        fun toggle(boxNo: Int): Boolean =
            data.remove(boxNo) || (data.size < MAX_BOOKMARKS && data.add(boxNo))

        fun remove(boxNo: Int) {
            data.remove(boxNo)
        }

        fun clear() {
            data.clear()
        }

        fun isFull(): Boolean = data.size >= MAX_BOOKMARKS
        fun has(boxNo: Int): Boolean = boxNo in data
        fun get(): List<Int> = data.toList()
    }
}